

diff --git a/xenmgr/XenMgr/Expose/VmDiskObject.hs b/xenmgr/XenMgr/Expose/VmDiskObject.hs
index efc59ea..fcc7b13 100644
--- a/xenmgr/XenMgr/Expose/VmDiskObject.hs
+++ b/xenmgr/XenMgr/Expose/VmDiskObject.hs
@@ -121,10 +121,9 @@ _GetBackendUuid :: ID -> Rpc String
 _GetBackendUuid id = fromMaybe "" . fmap show  <$> _get_field id diskBackendUuid
 
 _SetBackendUuid :: ID -> String -> Vm ()
-_SetBackendUuid id uuid = _modify_disk id $ \d -> d { diskBackendName = case uuid of
-                                                                         "" -> Nothing
-                                                                         _  -> Just uuid }
-
+_SetBackendUuid id uuid = _modify_disk id $ \d -> d { diskBackendUuid = case uuid of
+                                                                           "" -> Nothing
+                                                                           _  -> Just (fromString uuid) }
 _GetBackendName :: ID -> Rpc String
 _GetBackendName id = fromMaybe "" . fmap show  <$> _get_field id diskBackendName
 
